### PR TITLE
Split `map::process_fields_in_submap` into field processors

### PR DESCRIPTION
**Pull request by Aivean**
#### Summary

Performance "Split `map::process_fields_in_submap` into field processors"

#### Purpose of change

Second step towards fields performance (see #47032).

Currently `process_fields_in_submap` rechecks whether a particular field processing code is applicable to the field type for every existing field entry on the map.

Additionally, some of these checks iterate over all intensity levels to calculate a single boolean, that could be cached in the field type.

#### Describe the solution

The biggest change here is introduction of the "field processor":
```cpp
using FieldProcessorPtr = void( * )( const tripoint &p, field_entry &cur, field_proc_data &pd );
```
<details><summary>also see `field_proc_data`</summary>

```cpp
struct field_proc_data {
    scent_block &sblk;
    Character &player_character;
    const oter_id &om_ter;
    map &here;
    maptile &map_tile;
    field_type_id cur_fd_type_id;
    field_type const *cur_fd_type;
};
```
</details>

List of these processors is generated once per field type and then only relevant field processors are called in `process_fields_in_submap`.

Supplementary changes:
* cached `field_type::is_transparent` and `field_type::is_dangerous`
* removed lots of trivial boilerplate methods from `field_type` and `field_entry`
* consolidated map cache invalidation logic in `map::on_field_modified`

#### Describe alternatives you've considered

At first I thought about generating list of field processors on the fly, but current solution is better in many ways.

#### Testing

Field unit tests added in #47061.
Manual smoke test (pun intended).

#### Additional context

Fire processing speed in (hotel on fire) scenario improved ≈7%-10%.

Profile before:
image

image

After:
image

image

### CODE REVIEW

File: src/map_field.cpp [hunk 1]
```diff
@@ -436,483 +441,485 @@ void map::process_fields_in_submap( submap *const current_submap,
...
+{
+    // Upgrade field intensity
+    const field_intensity_level &int_level = cur.get_intensity_level();
+    if( int_level.intensity_upgrade_chance > 0 &&
+        int_level.intensity_upgrade_duration > 0_turns &&
+        calendar::once_every( int_level.intensity_upgrade_duration ) &&
+        one_in( int_level.intensity_upgrade_chance ) ) {
+        cur.mod_field_intensity( 1 );
+    }
+};
```
Comment (BrettDong) [inline src/map_field.cpp:498]:
```suggestion
}
```

File: src/map_field.cpp [hunk 2]
```diff
@@ -436,483 +441,485 @@ void map::process_fields_in_submap( submap *const current_submap,
...
+        cur.set_field_intensity( 0 );
+    }
 
-                // Apply radiation
-                if( cur.extra_radiation_max() > 0 ) {
-                    int extra_radiation = rng( cur.extra_radiation_min(), cur.extra_radiation_max() );
-                    adjust_radiation( p, extra_radiation );
-                }
+    // TODO: Allow spreading to the sides if age < 0 && intensity == 3
+};
```
Comment (BrettDong) [inline src/map_field.cpp:540]:
```suggestion
}
```

**Comment by int-ua** [comment]
Resetting Actions after the GitHub incident https://www.githubstatus.com/incidents/tf9v5jjmq2lg by closing and reopening.

File: src/map_field.cpp [hunk 3]
```diff
@@ -436,483 +441,481 @@ void map::process_fields_in_submap( submap *const current_submap,
...
+                tripoint newp = random_entry( valid );
+                pd.here.add_item_or_charges( newp, tmp );
+                if( pd.player_character.pos() == newp ) {
+                    add_msg( m_bad, _( "A %s hits you!" ), tmp.tname() );
+                    const bodypart_id hit = pd.player_character.get_random_body_part();
+                    pd.player_character.deal_damage( nullptr, hit, damage_instance( damage_type::BASH, 6 ) );
+                    pd.player_character.check_dead_state();
+                }
+
+                if( npc *const n = g->critter_at<npc>(newp ) ) {
```
Comment (BrettDong) [inline src/map_field.cpp:734]:
```suggestion
                if( npc *const n = g->critter_at<npc>( newp ) ) {
```

File: src/map_field.cpp [hunk 4]
```diff
@@ -436,483 +441,481 @@ void map::process_fields_in_submap( submap *const current_submap,
...
+                    // TODO: combine with player character code above
+                    const bodypart_id hit = pd.player_character.get_random_body_part();
+                    n->deal_damage(nullptr, hit, damage_instance(damage_type::BASH, 6 ) );
+                    add_msg_if_player_sees(newp, _( "A %1$s hits %2$s!" ), tmp.tname(), n->name );
```
Comment (BrettDong) [inline src/map_field.cpp:738]:
```suggestion
                    n->deal_damage( nullptr, hit, damage_instance( damage_type::BASH, 6 ) );
                    add_msg_if_player_sees( newp, _( "A %1$s hits %2$s!" ), tmp.tname(), n->name );
```

**Review by BrettDong — COMMENTED** [review]
`astyle` errors:

File: src/map_field.cpp (hunk 3 above)
Comment (Aivean) [reply src/map_field.cpp]: Thank you. I wonder how that happened, I even have git commit hook with astyle.

**Pull request merged**